### PR TITLE
Feature/simgan support block code

### PR DESCRIPTION
**Pull request by adachille**
https://github.com/ezCGP/ezCGP/issues/123

This pull starts to flesh out the block code for the simgan problem. Most notably, it splits the single block I was using into two blocks, one for the SimGAN refiner and one for the SimGAN discriminator. This PR also includes the adding of simple PyTorch operators, and a general BlockEvaluate class for building PyTorch graphs, which are slightly different than for Tensorflow.

### CODE REVIEW

File: problems/problem_tfkeras_transferlearning.py [hunk 1]
```diff
@@ -70,7 +70,7 @@ class Problem(ProblemDefinition_Abstract):
     '''
     def __init__(self):
         import tensorflow as tf
-        assert(len(tf.config.experimental.list_physical_devices('GPU'))>=1), "GPU NOT FOUND - ezCGP EXITING"
+        # assert(len(tf.config.experimental.list_physical_devices('GPU'))>=1), "GPU NOT FOUND - ezCGP EXITING"
```
Comment (roddtalebi) [inline problems/problem_tfkeras_transferlearning.py:73]: woops sorry

**Review by roddtalebi — CHANGES_REQUESTED** [review]
welp im impressed by how quickly you got into the swing of things . nice

File: codes/block_definitions/block_evaluate.py [hunk 2]
```diff
@@ -831,3 +835,185 @@ def evaluate(self,
             return
         
         block_material.output = [None, output] # TODO make sure it is a list
+
+class BlockEvaluate_PyTorch(BlockEvaluate_GraphAbstract):
+    @abstractmethod
```
[the remark this answers is not in the document]
Comment (adachille) [reply codes/block_definitions/block_evaluate.py:840]: Ah good, thanks for letting me know.

File: codes/block_definitions/block_evaluate.py [hunk 3]
```diff
@@ -344,6 +346,7 @@ def build_graph(self, block_material, block_def, datapair):
                                                   block_def,
                                                   [input_layer])[0]
 
+        import pdb; pdb.set_trace()
```
[the remark this answers is not in the document]
Comment (adachille) [reply codes/block_definitions/block_evaluate.py:349]: Good catch

File: codes/block_definitions/block_evaluate.py [hunk 4]
```diff
@@ -831,3 +835,185 @@ def evaluate(self,
...
+            else:
+                # main node. this is where we evaluate
+                function = block_material[node_index]["ftn"]
+                args = []
+                node_arg_indices = block_material[node_index]["args"]
+                for node_arg_index in node_arg_indices:
+                    args.append(block_material.args[node_arg_index].value)
+
+                ezLogging.debug("%s - Builing %i; Function: %s, Inputs: %s, Args: %s" % (block_material.id, node_index, function, inputs, args))
+                node = function(inputs[-1], *args)
```
[the remark this answers is not in the document]
Comment (adachille) [reply codes/block_definitions/block_evaluate.py:865]: Hm, this is something I would like to talk to you about more. I want to figure out a way to allow for more than 1 input, but I'm just not 100% sure what that looks like.

**Comment by adachille** [comment]
I changed the code so that every operator is a subclass of PyTorchLayerWrapper. This makes everything much much simpler, allows for multiple inputs (even if not all are utilized), and makes building graphs easy!

**Review by roddtalebi — APPROVED** [review]
i never really liked nn.Sequential for building a graph but I think i can see what is happening and how it works. you have my blessings.

**Pull request merged**